web/js/codeworld-mode.js

```diff
@@ -82,43 +116,91 @@ CodeMirror.defineMode('codeworld', (_config, modeConfig) => {
 
         if (stream.match(RE_DASHES)) {
             stream.skipToEnd();
-            return 'comment';
+            return ['comment'];
         }
 
-        if (stream.match(RE_QUAL)) return 'qualifier';
-        if (stream.match(RE_VARID) || stream.match(RE_VARSYM)) return 'variable';
-        if (stream.match(RE_CONID) || stream.match(RE_CONSYM)) return 'variable-2';
-        if (stream.match(RE_NUMBER)) return 'number';
-        if (stream.match(RE_CHAR) || stream.match(RE_STRING)) return 'string';
+        if (checkContextStarter(stream, state)) {
+            return ['keyword'];
+        }
+
+        if (state.scanState === NORMAL && stream.match(RE_IN)) {
+            // Cancel previous layout 'let' context
```

Comment:
You really need to verify the user is in a layout context for `let` here.  For example, imagine the user mistyped: `colored(solidCircle(3), lt c = dark(green) in translucent(c))`  The type of adding `lt` instead of `let` causes `in` to break the parentheses, and it's all chaos.

For that matter, what about this correct code?

    let { x = 5 } in x + 37

Here there is no layout context when you encounter the `in`.  I believe the current code will mistakenly pop some surrounding bracket or layout context.

Reply:
Okey, this is about upgrading the `layoutContext` to store more information.

Now `layoutContext` contain brackets and integers. Thinking about transform it to something like
```
[{type: 'bracket', contextStarterColumn: 0}
,{type: 'square', contextStarterColumn: 1}
,{type: 'curly', contextStarterColumn: 2}
,{type: 'where',  contextStarterColumn: 8, contextColumn: 14}
,{type: 'let', contextStarterColumn: 14, contextColumn: 18}]
```

Reply:
This looks like a good plan.  I'm not sure I understand what `contextStarterColumn` is about, but the rest seems right.